File: week 3/hash_substring/hash_substring.cpp
#include <iostream>
#include <string>
#include <vector>

using std::string;
typedef unsigned long long ull;

struct Data {
	string pattern, text;
};

Data read_input() {
	Data data;
	std::cin >> data.pattern >> data.text;
	return data;
}

void print_occurrences(const std::vector<int>& output) {
	for (size_t i = 0; i < output.size(); ++i)
		std::cout << output[i] << " ";
	std::cout << "\n";
}

std::vector<int> get_occurrences_naive(const Data& input) {
	const string& s = input.pattern, t = input.text;
	std::vector<int> ans;
	for (size_t i = 0; i + s.size() <= t.size(); ++i) {
		if (t.substr(i, s.size()) == s)
			ans.push_back(i);
	}
	return ans;
}

long long polyHash(const string& s, int prime, int x) {
	unsigned long long hash = 0;
	for (long long i = static_cast<long long> (s.size()) - 1; i >= 0; --i)
		hash = (hash * x + s[i]) % prime;
	return hash;
}

std::vector<long long> preComputeHashes(string text, int pLength, int p, int x) {
	int tLength = text.size();
	std::vector<long long> H(tLength - pLength + 1);
	string s = text.substr(tLength - pLength, pLength);
	H[tLength - pLength] = polyHash(s, p, x);
	long long y = 1;
	for (int i = 1; i <= pLength; ++i)
		y = y * x % p;
	for (int i = tLength - pLength - 1; i >= 0; --i) {
		long long preHash = x * H[i + 1] + text[i] - y * text[i + pLength];
		while (preHash < 0) {
			preHash += p;
		}
		H[i] = preHash % p;
	}
	return H;
}

bool areEqual(string s1, string s2){
	if (s1.size() != s2.size())
		return false;
	for (int i = 0; i < s1.size(); ++i) {
		if (s1[i] != s2[i])
			return false;
	}
	return true;
}

std::vector<int> get_occurrences(const Data& input) {
	int p = 1000000007;
	int x = rand() % (p-1) + 1;
	const string& s = input.pattern, t = input.text;
	std::vector<int> ans;

	long long pHash = polyHash(input.pattern, p, x);
	std::vector<long long> H = preComputeHashes(t, s.size(), p, x);
	for (size_t i = 0; i + s.size() <= t.size(); ++i) {
		if (pHash != H[i])
			continue;
		if (areEqual(t.substr(i, s.size()), s))
			ans.push_back(i);
	}
	return ans;
}

int main() {
	std::ios_base::sync_with_stdio(false);
	print_occurrences(get_occurrences(read_input()));
	return 0;
}

File: week 3/hash_substring/hash_substring.py
import random

def read_input():
    return (raw_input().rstrip(), raw_input().rstrip())

def print_occurrences(output):
    print(' '.join(map(str, output)))

def get_occurrences_naive(pattern, text):
    return [
        i 
        for i in range(len(text) - len(pattern) + 1) 
        if text[i:i + len(pattern)] == pattern
    ]

def poly_hash(s, prime, x):
    ans = 0
    for c in reversed(s):
    	ans = (ans * x + ord(c)) % prime
    return ans

def precompute_hashes(text, plength, p, x):
    H = [0] * (len(text) - plength + 1)
    s = text[-plength:]
    H[len(text)-plength] = poly_hash(s, p, x)
    y = 1
    for i in range(1, plength+1):
        y = (y * x) % p
    for i in reversed(range(len(text) - plength)):
    	prehash = x * H[i + 1] + ord(text[i]) - y * ord(text[i + plength])
    	while(prehash < 0):
    		prehash += p
        H[i] = prehash % p
    return H

def get_occurrences(pattern, text):
    p = 1000000007
    x = random.randint(1, p)
    tlength = len(text)
    plength = len(pattern)
    phash = poly_hash(pattern, p, x)
    H = precompute_hashes(text, plength, p, x)
    return [
        i 
        for i in range(len(text) - len(pattern) + 1) 
        if phash == H[i] and text[i:i + len(pattern)] == pattern
    ]

if __name__ == '__main__':
    print_occurrences(get_occurrences(*read_input()))